CONTRIBUTING.md

```diff
@@ -604,35 +604,222 @@ about the error diagnosis below, if it doesn't work.
+UPPERCASE, which gives rise to the possibility of clobbering a legitimate 
+global var.  This can have results that are 
+[potentially hilarious](https://stackoverflow.com/q/28310594), or potentially bad.
+
+For that reason, UPPERCASE variable names are strongly discouraged outside of
+the global scope.
+
+Instead, use lowercase, with underscores to separate words i.e. `snake_case`.
+
+Try, also, to use meaningful names, this is meaningless:
+
+```
+for f in df; do
+  ...
+```
+
+Whereas this is better:
+
+```
+for file_system in $(df); do
+```
+
+*Exception: C-Style `for` loops :)*
+
+#### Function / Local
+
+`bash` and others allow you to define variables as local within a function e.g.
 
 ```
-  sudo docker run --rm -v "$(pwd):/sh" -w /sh peterdavehello/shfmt shfmt -i 4 -ci -w agents/check_mk_agent.linux
+get_api_user() {
+  local username=Shelly
+}
 ```
 
+Unfortunately this is not portable.  So our practice is to prepend any variables
+within a function with an underscore.  We also `unset` the variable immediately
+prior to the function closure.  For example:
+
+```
+get_api_user() {
+  _username=Shelly
+  printf -- '%s\n' "${_username}"
+  unset -v _username
+}
+```
+
+#### Curly braces?
+
+Curly braces are used on `${arrays[@]}` and `${variable/modif/ications}`.
+To maintain consistency, you should use curly braces on normal variables too.
+
+Curly braces around variables improves readability when syntax colouring is not
+available.  `${this_variable}` stands out within this block of text.
+
+They also allow us to better format outputs e.g.
+
+```
+▓▒░$ time_metric=5
+▓▒░$ echo "$time_metricmins"
+
+▓▒░$ echo "${time_metric}mins"
+5mins
+```
+
+In the first example, there is no such variable as `$timemetricmins`.
+
+So curly braces explicitly tell the shell interpreter where the variable name
+boundaries are.
+
+*Exception: When you're in an arithmetic context e.g. `(( time_metric < 10 ))`*
+
+#### Re-cap:
+
+* **Global / Environment**: `${MK_UPPERCASE}`
+* **Constants**: `${MK_UPPERCASE}` set to `readonly`
+* **Script**: `${meaningful_snake_case}`
+* **Function / Local** `${_underscore_prepended_snake_case}` with `unset -v`
+
+### Linting and formatting
+
+Use [shellcheck](https://www.shellcheck.net/) for your changes before
+submitting patches.
+
+The best results are achieved with
+[a direct integration](https://github.com/koalaman/shellcheck#user-content-in-your-editor)
+into your preferred editor, so that you are immediately informed about
+possible problems during programming.
+
+The various agent scripts are not clean at the moment, but we aim to clean
+this up in the near future.
+
+Do note that while Shellcheck is an excellent tool, it's not perfect.
+Sometimes it alerts on code that may actually be desired.  In this scenario,
+you must use a `disable` directive with a comment that justifies your reason
+for this.  Preferably this will be a git commit hash where the reasoning is
+given e.g.
+
+```
+# See commit abcd3456
```

Comment:
Better than referring to a commit is to directly add the reason to the comment line(s) above.

Reply:
Reworded in the latest commit(s)